builtin/packer/config_sourcer.go

```diff
@@ -0,0 +1,235 @@
+func (cs *ConfigSourcer) StopFunc() interface{} {
+	return cs.stop
+}
+
+func (cs *ConfigSourcer) read(
+	ctx context.Context,
+	log hclog.Logger,
+	reqs []*component.ConfigRequest,
+) ([]*pb.ConfigSource_Value, error) {
+
+	// If the user has explicitly set the client ID and secret for the config
+	// sourcer, we use that. Otherwise, we use environment variables.
+	opts := hcpconfig.FromEnv()
+	if cs.config.ClientId != "" && cs.config.ClientSecret != "" {
+		opts = hcpconfig.WithClientCredentials(cs.config.ClientId, cs.config.ClientSecret)
+	}
+	hcpConfig, err := hcpconfig.NewHCPConfig(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	hcpClient, err := httpclient.New(httpclient.Config{
+		HCPConfig: hcpConfig,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	hcpPackerClient := packer.New(hcpClient, nil)
+	channelParams := packer.NewPackerServiceGetChannelParams()
+	channelParams.LocationOrganizationID = cs.config.OrganizationId
+	channelParams.LocationProjectID = cs.config.ProjectId
+
+	var results []*pb.ConfigSource_Value
+	for _, req := range reqs {
+		result := &pb.ConfigSource_Value{Name: req.Name}
+		results = append(results, result)
+
+		var packerConfig reqConfig
+		// We serialize the config sourcer settings to the reqConfig struct.
+		if err = mapstructure.WeakDecode(req.Config, &packerConfig); err != nil {
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, err.Error()).Proto(),
+			}
+			return nil, err
+		}
+		channelParams.BucketSlug = packerConfig.Bucket
+		channelParams.Slug = packerConfig.Channel
+
+		// An HCP Packer channel points to a single iteration of a bucket.
+		channel, err := hcpPackerClient.PackerServiceGetChannel(channelParams, nil)
+		if err != nil {
+			return nil, err
+		}
+		log.Debug("Retrieved HCP Packer channel.")
+		iteration := channel.Payload.Channel.Iteration
+
+		// An iteration can have multiple builds, so we check for the first build
+		// with the matching cloud provider and region.
+		for _, build := range iteration.Builds {
+			if build.CloudProvider == packerConfig.Cloud {
+				log.Debug("Found build with matching cloud provider.")
+				for _, image := range build.Images {
+					if image.Region == packerConfig.Region {
+						log.Debug("Found image with matching region.")
+						result.Result = &pb.ConfigSource_Value_Value{
+							// The ImageID is the Cloud Image ID or URL string
+							// identifying this image for the builder that built it,
+							// so this is returned to Waypoint.
+							Value: image.ImageID,
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return results, nil
+}
+
+func (cs *ConfigSourcer) stop() error {
+	return nil
+}
+
+func (cs *ConfigSourcer) Documentation() (*docs.Documentation, error) {
+	doc, err := docs.New(
+		docs.FromConfig(&sourceConfig{}),
+		docs.RequestFromStruct(&reqConfig{}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	doc.Description("Read machine image information from HCP Packer.")
```

Comment:
We should probably have a bit more information here, like for example that only container images are valid, etc.

Reply:
I'm hesitant to point out that only container images will work, since Packer supports more than just containers. Our plugins do work almost exclusively with containers, but since plugins don't _have_ to, and since we have the EC2 plugin, I think it's okay to leave it at "image". However, I will at least add some color to this description! 😃